### PR TITLE
Add EarlyStopping

### PR DESCRIPTION
**Pull request by LiuChiachi**
### PR types

New features
### PR changes

APIs
### Describe

Add `Callback` subclass `EarlyStopping`

### CODE REVIEW

File: python/paddle/hapi/callbacks.py [hunk 1]
```diff
@@ -485,6 +494,158 @@ def on_train_end(self, logs=None):
...
+        self.best_weights = None
+        self.stopped_epoch = 0
+        self.save_best_model = save_best_model
+        self.save_dir = None  # `save_dir` is get from `config_callbacks`
+        if mode not in ['auto', 'min', 'max']:
+            warnings.warn('EarlyStopping mode %s is unknown, '
+                          'fallback to auto mode.' % mode)
+            mode = 'auto'
+        if mode == 'min' or 'loss' in self.monitor:
+            self.monitor_op = np.less
```
Comment (guoshengCS) [inline python/paddle/hapi/callbacks.py:687]: 这里的处理是否有问题，如果显式设置了`mode = 'max'`，感觉这里应该还是需要按`mode == 'max'`走而不是通过`self.monitor`的name来推断吧，推断是否应该只发生在`auto`的情况，另外`loss`是否一定就和min对应呢。也在文档中加上推断的规则说明吧。

File: python/paddle/hapi/callbacks.py [hunk 2]
```diff
@@ -485,6 +494,158 @@ def on_train_end(self, logs=None):
...
+class EarlyStopping(Callback):
+    """
+    Stop training when the given monitor stopped improving.
+    Args:
+        monitor(str): Quantity to be monitored. Default: 'loss'.
+        mode(str|None): Mode should be one of `auto`, `min` or `max`. In `min`
+            mode, training will stop until monitored quantity has stopped
+            increasing. In 'max' mode, training will stop until monitored
+            quantity has stopped decreasing. In 'auto' mode, exact mode can be
+            inferred by the name of monitor. Default: 'auto'.
```
Comment (guoshengCS) [inline python/paddle/hapi/callbacks.py]: 也在文档中加上推断的规则说明吧。

File: python/paddle/hapi/callbacks.py [hunk 3]
```diff
@@ -485,6 +494,158 @@ def on_train_end(self, logs=None):
...
+
+            inputs = [InputSpec([None, 1, 28, 28], 'float32', 'x')]
+            labels = [InputSpec([None, 1], 'int64', 'label')]
+
+            model = Model(net, inputs=inputs, labels=labels)
+            model.prepare(optim,
+                        loss=CrossEntropyLoss(reduction="sum"),
+                        metrics=[Accuracy()])
+            callbacks = paddle.callbacks.EarlyStopping('loss',
+                                                    mode='min',
```
Comment (guoshengCS) [inline python/paddle/hapi/callbacks.py]: 这里的格式最好调整下。

File: python/paddle/hapi/callbacks.py [hunk 4]
```diff
@@ -485,6 +494,158 @@ def on_train_end(self, logs=None):
             self.model.save(path)
 
 
+class EarlyStopping(Callback):
+    """
+    Stop training when the given monitor stopped improving.
```
Comment (guoshengCS) [inline python/paddle/hapi/callbacks.py]: 是否要说明这里只对eval生效

File: python/paddle/hapi/callbacks.py [hunk 5]
```diff
@@ -485,6 +494,158 @@ def on_train_end(self, logs=None):
...
+    Args:
+        monitor(str): Quantity to be monitored. Default: 'loss'.
+        mode(str|None): Mode should be one of `auto`, `min` or `max`. In `min`
+            mode, training will stop until monitored quantity has stopped
+            increasing. In 'max' mode, training will stop until monitored
```
Comment (guoshengCS) [inline python/paddle/hapi/callbacks.py]: 这里是否反了，`min`是对应decreasing，`max`是increasing吧

File: python/paddle/hapi/callbacks.py (hunk 2 above)
Comment (LiuChiachi) [reply python/paddle/hapi/callbacks.py]: Done. Thanks

File: python/paddle/hapi/callbacks.py (hunk 1 above)
Comment (LiuChiachi) [reply python/paddle/hapi/callbacks.py:687]: 感谢。已修改为对`monitor`中含有`"acc"`的推断为`"max"`，其它情况推断为`"min"`，并在文档中做出了说明。

File: python/paddle/hapi/callbacks.py (hunk 3 above)
Comment (LiuChiachi) [reply python/paddle/hapi/callbacks.py]: Done. Thanks

File: python/paddle/hapi/callbacks.py (hunk 4 above)
Comment (LiuChiachi) [reply python/paddle/hapi/callbacks.py]: Done. Thanks

File: python/paddle/hapi/callbacks.py (hunk 5 above)
Comment (LiuChiachi) [reply python/paddle/hapi/callbacks.py]: 感谢，已修改

File: python/paddle/hapi/callbacks.py [hunk 6]
```diff
@@ -485,6 +494,162 @@ def on_train_end(self, logs=None):
...
+
+            device = paddle.set_device('cpu')
+            sample_num = 200
+            save_dir = './best_model_checkpoint'
+            transform = T.Compose([T.Transpose(), T.Normalize([127.5], [127.5])])
+            train_dataset = MNIST(mode='train', transform=transform)
+            val_dataset = MNIST(mode='test', transform=transform)
+
+            train_loader = paddle.io.DataLoader(
+                train_dataset, places=device, return_list=True, batch_size=64)
```
Comment (saxon-zh) [inline python/paddle/hapi/callbacks.py]: dataloader现在可以不用写device

Comment (saxon-zh) [reply python/paddle/hapi/callbacks.py]: 另外这里示例是不是不需要单独再写下data loader

**Review by XiaoguangHu01 — DISMISSED** [review]
LGTM

File: python/paddle/hapi/callbacks.py (hunk 6 above)
Comment (LiuChiachi) [reply python/paddle/hapi/callbacks.py]: Done. Thanks.

**Review by XiaoguangHu01 — DISMISSED** [review]
LGTM

**Review by saxon-zh — DISMISSED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by Heeenrrry — APPROVED** [review]
